### ISSUE TITLE
Confused type inference with function overload

### ISSUE THREAD

**Issue by inad9300** [microsoft/TypeScript#31578]
**TypeScript Version:** 3.5.0-dev.20190523

**Search Terms:** broken type inference / checker in function overloading with generics...

**Code**

I find it quite difficult to explain the problem in words. In essence, TypeScript is somehow not inferring the generic types correctly. I believe it is best to just look at the code:

```ts
interface HTMLElementChildrenMap extends Record<keyof HTMLElementTagNameMap, void | Element[]> {
    hr:  void
    table: (HTMLElementTagNameMap['thead'] | HTMLElementTagNameMap['tbody'] | HTMLElementTagNameMap['tfoot'])[]
}

function h<T extends keyof HTMLElementTagNameMap, E extends HTMLElementTagNameMap[T]>(tag: T, children?: HTMLElementChildrenMap[T]): E
function h<T extends keyof HTMLElementTagNameMap, E extends HTMLElementTagNameMap[T]>(elem: E, children?: HTMLElementChildrenMap[T]): E
function h(tagOrElem: keyof HTMLElementTagNameMap | HTMLElement): HTMLElement {
    return null
}

// Errors caught (expected behavior)
h('hr', [])
h('table', [document.createElement('button')]) 

// Errors not caught
h(document.createElement('hr'), [])
h(document.createElement('table'), [document.createElement('button')])
```

**Playground Link:** [here](https://www.typescriptlang.org/play/#src=interface%20HTMLElementChildrenMap%20extends%20Record<keyof%20HTMLElementTagNameMap%2C%20void%20%7C%20Element%5B%5D>%20%7B%0D%0A%20%20%20%20hr%3A%20%20void%0D%0A%20%20%20%20table%3A%20(HTMLElementTagNameMap%5B'thead'%5D%20%7C%20HTMLElementTagNameMap%5B'tbody'%5D%20%7C%20HTMLElementTagNameMap%5B'tfoot'%5D)%5B%5D%0D%0A%7D%0D%0A%0D%0Afunction%20h<T%20extends%20keyof%20HTMLElementTagNameMap%2C%20E%20extends%20HTMLElementTagNameMap%5BT%5D>(tag%3A%20T%2C%20children%3F%3A%20HTMLElementChildrenMap%5BT%5D)%3A%20E%0D%0Afunction%20h<T%20extends%20keyof%20HTMLElementTagNameMap%2C%20E%20extends%20HTMLElementTagNameMap%5BT%5D>(elem%3A%20E%2C%20children%3F%3A%20HTMLElementChildrenMap%5BT%5D)%3A%20E%0D%0Afunction%20h(tagOrElem%3A%20keyof%20HTMLElementTagNameMap%20%7C%20HTMLElement)%3A%20HTMLElement%20%7B%0D%0A%20%20%20%20return%20null%0D%0A%7D%0D%0A%0D%0Ah('hr'%2C%20%5B%5D)%0D%0Ah('table'%2C%20%5Bdocument.createElement('button')%5D)%0D%0A%0D%0Ah(document.createElement('hr')%2C%20%5B%5D)%0D%0Ah(document.createElement('table')%2C%20%5Bdocument.createElement('button')%5D))

A secondary problem is that the error are very much not helpful in the snippet above. They become useful when the order of the overloads is reversed:

```ts
interface HTMLElementChildrenMap extends Record<keyof HTMLElementTagNameMap, void | Element[]> {
    hr:  void
    table: (HTMLElementTagNameMap['thead'] | HTMLElementTagNameMap['tbody'] | HTMLElementTagNameMap['tfoot'])[]
}

function h<T extends keyof HTMLElementTagNameMap, E extends HTMLElementTagNameMap[T]>(elem: E, children?: HTMLElementChildrenMap[T]): E
function h<T extends keyof HTMLElementTagNameMap, E extends HTMLElementTagNameMap[T]>(tag: T, children?: HTMLElementChildrenMap[T]): E
function h(tagOrElem: keyof HTMLElementTagNameMap | HTMLElement): HTMLElement {
    return null
}

h('hr', [])
h('table', [document.createElement('button')])
```

**Playground Link:** [here](https://www.typescriptlang.org/play/#src=interface%20HTMLElementChildrenMap%20extends%20Record<keyof%20HTMLElementTagNameMap%2C%20void%20%7C%20Element%5B%5D>%20%7B%0D%0A%20%20%20%20hr%3A%20%20void%0D%0A%20%20%20%20table%3A%20(HTMLElementTagNameMap%5B'thead'%5D%20%7C%20HTMLElementTagNameMap%5B'tbody'%5D%20%7C%20HTMLElementTagNameMap%5B'tfoot'%5D)%5B%5D%0D%0A%7D%0D%0A%0D%0Afunction%20h<T%20extends%20keyof%20HTMLElementTagNameMap%2C%20E%20extends%20HTMLElementTagNameMap%5BT%5D>(elem%3A%20E%2C%20children%3F%3A%20HTMLElementChildrenMap%5BT%5D)%3A%20E%0D%0Afunction%20h<T%20extends%20keyof%20HTMLElementTagNameMap%2C%20E%20extends%20HTMLElementTagNameMap%5BT%5D>(tag%3A%20T%2C%20children%3F%3A%20HTMLElementChildrenMap%5BT%5D)%3A%20E%0D%0Afunction%20h(tagOrElem%3A%20keyof%20HTMLElementTagNameMap%20%7C%20HTMLElement)%3A%20HTMLElement%20%7B%0D%0A%20%20%20%20return%20null%0D%0A%7D%0D%0A%0D%0Ah('hr'%2C%20%5B%5D)%0D%0Ah('table'%2C%20%5Bdocument.createElement('button')%5D))

**Related Issues:** https://github.com/microsoft/TypeScript/issues/27972, https://github.com/microsoft/TypeScript/issues/20396, https://github.com/microsoft/TypeScript/issues/10645, https://github.com/microsoft/TypeScript/issues/21707

**Comment by MartinJohns** [comment]
In your second overload there is nothing that `T` can be inferred from. As a result `T` will fall back to being `keyof HTMLElementTagNameMap`, and `HTMLElementChildrenMap[T]` will be a union of all possible combinations. One of those combinations is provided by your `table` property, which accepts an array of type `HTMLElementTagNameMap['thead'] | HTMLElementTagNameMap['tbody'] | HTMLElementTagNameMap['tfoot']` - the empty array `[]` matches this constraint.

You can see this when hovering over the `h` in your method invocation and check the tooltip.

**Comment by inad9300** [comment]
Thanks for your reply. In my mind, `T` can be inferred from `E` quite easily. I was expecting TypeScript to be able to figure this out as well, as from my experience TypeScript's inference power is great.

How can I type this, then, so that there isn't a need to explicitly specify the generic types? Is there a way to somehow have a reversed version of `HTMLElementTagNameMap` that goes from interface to tag?

**Comment by inad9300** [comment]
Could perhaps someone tell me at least if this problem is actually solvable? It might be more obvious for you than it is for me... That would be really helpful already.

**Comment by inad9300** [comment]
I'm pretty convinced that this problem indeed can't be solved with current TypeScript features. That is because `HTMLElementTagNameMap` maps multiple tags to the same `HTMLElement` interface, therefore making it impossible for the compiler to identify whether (e.g.) `<abbr>` or `<em>` has been provided. Even then, I believe nominal types would be necessary to be able to write a function like the one above.

Another version of the problem which is also not solvable at the moment (and which I would like to see TypeScript do at some point) is improving the type safety of existing HTML interfaces by restricting the way HTML hierarchies can be formed. For example:

```ts
interface HTMLTableElement {
    appendChild<T extends HTMLElementTagNameMap['thead'] | HTMLElementTagNameMap['tbody'] | HTMLElementTagNameMap['tfoot']>(newChild: T): T
}
```

This could be solved by giving specific names to all HTML element interfaces, e.g. `HTMLEmElement` for `<em>` and `HTMLAbbrElement` for `<abbr>`. If TypeScript included these declarations, even if they were empty at first, we could then extend them as shown above. My question then being: is this something you would consider doing? (Should I open a new issue?)

**Comment by inad9300** [comment]
I came across another example which seems to show that type inference is indeed not working properly, which I would like to report as a bug:

```ts
interface A {
    __d: 'A'
}

interface B {
    __d: 'B'
}

interface HtmlTypes {
    a: A
    b: B
}

interface HtmlChildren {
    a: void
    b: B[]
}

function h<T extends keyof HtmlTypes, E extends HtmlTypes[T]>(e?: E, c?: HtmlChildren[T]) {}

h(null as A, []) // Type inference not working: no errors reported.
h<'a', A>(null, []) // Explicit types given: errors correctly reported.
```

[Link to Playground.](https://www.typescriptlang.org/play/#src=interface%20A%20%7B%0A%20%20%20%20__d%3A%20'A'%0A%7D%0A%0Ainterface%20B%20%7B%0A%20%20%20%20__d%3A%20'B'%0A%7D%0A%0Ainterface%20HtmlTypes%20%7B%0A%20%20%20%20a%3A%20A%0A%20%20%20%20b%3A%20B%0A%7D%0A%0Ainterface%20HtmlChildren%20%7B%0A%20%20%20%20a%3A%20void%0A%20%20%20%20b%3A%20B%5B%5D%0A%7D%0A%0Afunction%20h%3CT%20extends%20keyof%20HtmlTypes%2C%20E%20extends%20HtmlTypes%5BT%5D%3E(e%3F%3A%20E%2C%20c%3F%3A%20HtmlChildren%5BT%5D)%20%7B%7D%0A%0Ah(null%20as%20A%2C%20%5B%5D)%20%2F%2F%20Type%20inference%20not%20working%3A%20no%20errors%20reported.%0Ah%3C'a'%2C%20A%3E(null%2C%20%5B%5D)%20%2F%2F%20Explicit%20types%20given%3A%20errors%20correctly%20reported.%0A)

I was trying to introduce a discriminator field to aid TypeScript, but it doesn't seem to be working as expected. Shouldn't this be possible, in the same way that the following is?

```ts
function h2<T extends keyof HtmlTypes, E extends HtmlTypes[T]>(t?: T, c?: HtmlChildren[T]) {}

h2('a', []) // Type inference working: errors correctly reported.
```

**Comment by inad9300** [comment]
Thanks, but no thanks, @typescript-bot.

Could you please re-open this issue until someone has time to look at it, @RyanCavanaugh? Thank you.
